## Chapter Five Results

## <font color='green'>Application: Calculate the forward zero prices for the bonds</font>



<div style="background-color:LightGray;
    border-left: 12px solid green;
    font-family: 'Garamond', serif;
    font-size: 17px;
    line-height: 1.5;
    padding: 15px">
<br>
Settlement date determined so that all payments occur at six-month candence.
<br>


|Maturity|Coupon|
|-------|-------|
|&nbsp;&nbsp;&nbsp;December 31$^{st}$ 2029|&nbsp;&nbsp;&nbsp;4.375|
|&nbsp;&nbsp;&nbsp;October 31$^{st}$ 2029|&nbsp;&nbsp;&nbsp;4.125|


1. **Import Required Modules:**  
   * Built-in modules  
   * Pandas (install if necessary)  
   * Custom modules:  
     * <font color='green'>accrued_interest</font>  
     * <font color='green'>create_payoff_matrix</font>  
     * <font color='green'>bond_pay_data</font>  
2. **Download Data:**  
   * Retrieve the <font color='green'>bond_data</font> file from DropBox.  
3. **Perform Calculations:**  
   * Calculate zero prices.  
   * Calculate normalizing zero prices  
   * Calculate forward zero\_prices.


</div>

### Required modules

In [2]:
# Import OS to interact with local computer operating system
import sys
import requests
from types import ModuleType
# Import the datetime and date classes from the datetime module
from datetime import datetime, date

# Import the Pandas library for data manipulation and analysis, aliased as pd.
try:
    import pandas as pd
except:
    !pip install pandas
    import pandas as pd

# Import the NumPy library for data manipulation and analysis, aliased as np.
try:
    import numpy as np
except:
    !pip install numpy
    import numpy as np

In [3]:
# Define the URL of the Python module to be downloaded from Dropbox.
# The 'dl=1' parameter in the URL forces a direct download of the file content.
url= 'https://www.dropbox.com/scl/fi/4y5hjxlfphh1ngvbgo77q/\
module_-basic_concepts_fixed_income.py?rlkey=6oxi7mgka42veaat79hcv8boz&st=87sztshr&dl=1'
module_name='basic_concepts_fixed_income'
# Send an HTTP GET request to the URL and store the server's response.
try:
  response=requests.get(url)
  # Raise an exception for bad status codes (like 404 Not Found)
  response.raise_for_status()
  module= ModuleType(module_name)
  #Code contained in response.text executed
  exec(response.text, module.__dict__)
  # Module added to sys
  sys.modules[module_name]=module
except requests.exceptions.RequestException as e:
    print(f"❌ Error: Could not fetch module from URL. {e}")
except Exception as e:
    print(f"❌ Error: Failed to execute or import the module. {e}")

# Now that 'basic_concepts_fixed_income' exists in the notebook, import the specific functions
from basic_concepts_fixed_income import (accrued_interest,
                                         create_payoff_matrix,
                                         bond_pay_data)

### Download bond data

In [5]:
#The full file path.
url='https://www.dropbox.com/scl/fi/lgnaj41bt8o9sv5a63rr1/\
bond_data_jan21_2025.xlsx?rlkey=twjzkcqo0g2ahvot78518ti4x&st=ihc5feog&dl=1'
print(f"Attempting to load data from: {url}")

#Load the data from Excel, using the first column as the index.
try:
    bond_data = pd.read_excel(url, index_col='Maturity Date',sheet_name='Fidelity Data')

    # Display the first and last 5 rows of the loaded DataFrame to verify it worked.
    display(bond_data)

except FileNotFoundError:
    print("\nERROR: File not found.")
    print("Please check that the 'folder' and 'file' variables are spelled correctly'\
' and that the file exists in that location.")

Attempting to load data from: https://www.dropbox.com/scl/fi/lgnaj41bt8o9sv5a63rr1/bond_data_jan21_2025.xlsx?rlkey=twjzkcqo0g2ahvot78518ti4x&st=ihc5feog&dl=1


Description  Coupon  \
Maturity Date                                                              
2025-01-28     UNITED STATES TREAS BILLS ZERO CPN       0.000...   0.000   
2025-01-30     UNITED STATES TREAS BILLS ZERO CPN       0.000...   0.000   
2025-01-31     UNITED STATES TREAS SER U-2025          1.3750...   1.375   
2025-01-31     UNITED STATES TREAS SER AW-2025         4.1250...   4.125   
2025-01-31     UNITED STATES TREAS SER G-2025          2.5000...   2.500   
...                                                          ...     ...   
2029-11-30     UNITED STATES TREAS SER AG-2029         4.1250...   4.125   
2029-11-30     UNITED STATES TREAS SER S-2029          3.8750...   3.875   
2029-12-31     UNITED STATES TREAS SER T-2029          3.8750...   3.875   
2029-12-31     UNITED STATES TREAS SER AH-2029         4.3750...   4.375   
2030-01-31     UNITED STATES TREAS SER G-2030          3.5000...   3.500   

               Price Bid  Price Ask  Bid Size  Ask Size  
Maturity Date                                            
2025-01-28        99.929     99.930    100000    100000  
2025-01-30        99.906     99.907     40000     40000  
2025-01-31        99.921     99.934     60000     60000  
2025-01-31        99.988     99.997     60000     60000  
2025-01-31        99.953     99.965     60000     60000  
...                  ...        ...       ...       ...  
2029-11-30        98.910     98.914    100000    100000  
2029-11-30        97.763     97.782     40000     40000  
2029-12-31        97.734     97.738     65000     65000  
2029-12-31        99.988     99.989      7000      7000  
2030-01-31        95.994     96.011     16000     16000  

[304 rows x 6 columns]

### Perform calculations

#### zero prices

In [6]:
# the settlement date for the data is Jan 31, 2025
settlement=date(2025,1,21)
# apply the accrued_interest function to the DataFrame
bond_data['Accrued'] = bond_data.apply(
    lambda x: accrued_interest(
        maturity=x.name, # Assuming index is maturity
        coupon=x['Coupon'],
        settlement=settlement,
        freq=2,
        day_type='Actual/Actual'
    ), axis=1
)

# Calculate transaction prices
transaction_prices=((bond_data['Price Bid']+bond_data['Price Ask'])/2+
                    bond_data['Accrued']).to_numpy()

# Create payoff matrix
payoffs_304,column_dates_304=create_payoff_matrix(bond_data,settlement=settlement,freq=2)

# Estimate the present value factors with least squares
zero_prices_304,ss2_304,rank_304,colInfo_304=np.linalg.lstsq(payoffs_304,transaction_prices)

# Maturity of the bonds in years
maturity_years=[(mat_date-settlement).days/365.25 for mat_date in column_dates_304]


# Continuously compounded annualized spot rates
spot_rates_304=-np.log(zero_prices_304)/np.array(maturity_years)

# Continuously compounded annualized forward rates
forward_rates_304=-np.diff(np.log(zero_prices_304))/np.diff(maturity_years)
forward_rates_304=np.insert(forward_rates_304,0,np.nan)

# Create a DataFrame of least squares present value estimates and display results
# First a dictionary
bonds_304={'Zero Prices: 304':zero_prices_304,
           'Spot Rates:304':spot_rates_304,
           'Forward Rates: 304':forward_rates_304}

# Then a DataFrame
df_304=pd.DataFrame(bonds_304,index=column_dates_304)
display(df_304)

# Create a DataFrame of summary statistics and display results
# First a dictionary
summary_304={'Average Sum Squared Errors':ss2_304/len(bond_data),
             'Payoff Rank':rank_304,
             'Column Max Info':np.max(colInfo_304),
             'Column Min Info':np.min(colInfo_304)}
# Then a DataFrame
df_summary_304=pd.DataFrame(summary_304,index=['Summary Statistics']).T
display(df_summary_304)

Zero Prices: 304  Spot Rates:304  Forward Rates: 304
2025-01-28          0.999295        0.036799                 NaN
2025-01-30          0.999065        0.037963            0.042038
2025-01-31          0.998886        0.040711            0.065439
2025-02-04          0.998475        0.039817            0.037581
2025-02-06          0.998240        0.040213            0.042987
...                      ...             ...                 ...
2029-10-31          0.812875        0.043390            0.043360
2029-11-15          0.809539        0.043874            0.100158
2029-11-30          0.809936        0.043402           -0.011943
2029-12-31          0.806965        0.043400            0.043302
2030-01-31          0.803448        0.043536            0.051455

[149 rows x 3 columns]

Summary Statistics
Average Sum Squared Errors            0.011753
Payoff Rank                         149.000000
Column Max Info                     267.431046
Column Min Info                     100.000000

## Use the first payment date as an assumed settlement date

To maintain alignment for semi-annual (six-month) par yield calculations, we use the first coupon payment date as the assumed settlement date. By asigning the settlement date as the first payment date the resulting series skips the first date and uses all the remaining dates that are each 6 months apart. Using our forward zero prices allows us to  estimate par yields for equal-sized six-month period.

In [7]:
# Use bond_pay_data function to return the full payment date array
settlement=date(2025,1,21)

# a value greater than zero is reqiiored
coupon=0.01

# get pay_data for both bonds
pay_data_dec=bond_pay_data(date(2029,12,31),0.01,settlement=settlement,freq=2)
pay_data_oct=bond_pay_data(date(2029,10,31),0.01,settlement=settlement,freq=2)

## Forward zero prices
The first payment date is the assumed settlement date and used to normalize forward zero prices.

In [8]:
# first payment date
normalize_dec=df_304['Zero Prices: 304'].loc[pay_data_dec[0][0]]
normalize_oct=df_304['Zero Prices: 304'].loc[pay_data_oct[0][0]]

# dates of forward zero prices start at second payment date
forward_zero_prices=[df_304['Zero Prices: 304'].loc[pay_data_dec[0][1:]]/normalize_dec,
                     df_304['Zero Prices: 304'].loc[pay_data_oct[0][1:]]/normalize_oct]
display(forward_zero_prices)

[2025-12-31    0.979271
 2026-06-30    0.958857
 2026-12-31    0.938939
 2027-06-30    0.918966
 2028-01-03    0.898935
 2028-06-30    0.879044
 2029-01-02    0.859154
 2029-07-02    0.840161
 2029-12-31    0.822042
 Name: Zero Prices: 304, dtype: float64,
 2025-10-31    0.979291
 2026-04-30    0.958846
 2026-11-02    0.938948
 2027-04-30    0.919132
 2027-11-01    0.899442
 2028-05-01    0.879359
 2028-10-31    0.859813
 2029-04-30    0.840726
 2029-10-31    0.822358
 Name: Zero Prices: 304, dtype: float64]

## <span style="text-align: left; color:green; font-family: 'Franklin Gothic Medium', sans-serif; margin-top: 1.0em; margin-bottom: 0em; font-style: italic;">Chapter Exercise</span>


<div style="background-color:LightGray;
    border-left: 12px solid green;
    font-family: 'Garamond', serif;
    font-size: 17px;
    line-height: 1.5;
    padding: 15px">
<br>

Calculate the par yields for the following bonds:


|Maturity|Coupon|
|-------|-------|
|&nbsp;&nbsp;&nbsp;December 31$^{st}$ 2029|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.375|
|&nbsp;&nbsp;&nbsp;October 31$^{th}$ 2029|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.125|
|&nbsp;&nbsp;&nbsp;January 31$^{th}$ 2030|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.50|$

1. Calculate forward zero prices for January 31$^{st}$ 2030  
2. Append January 31$^{st}$ 2030 forward prices to forward zeros calculated with the application.
3. Calculate par yields for each element of forward zeros  

   * <font color='green'>Dec_29</font>   
   * <font color='green'>Oct_29</font>   
   * <font color='green'>Jan_30</font>
4. Combine the par yield arrays with Pandas <font color='green'>concat</font> method and sort with <font color='green'>sort_index</font> method of Pandas
</div>

### Calculate forward zero prices for January 31$^{st}$ 2030 and append to forward_zeros


In [ ]:
settlement=date(2025,1,21)
# maturity is from the bond, any number greater than zero works for coupon
coupon=0.01
maturity=date(2030,1,31)

pay_data_jan=bond_pay_data(maturity,coupon,settlement=settlement,freq=2)

normalize_jan=df_304['Zero Prices: 304'].loc[pay_data_jan[0][0]]

forward_zero_jan=df_304['Zero Prices: 304'].loc[pay_data_jan[0][1:]]/normalize_jan

# forward zero prices calculated
forward_zero_prices.append(forward_zero_jan)

# forward zero prices for the three dates
display(forward_zero_prices)

[2025-12-31    0.979271
 2026-06-30    0.958857
 2026-12-31    0.938939
 2027-06-30    0.918966
 2028-01-03    0.898935
 2028-06-30    0.879044
 2029-01-02    0.859154
 2029-07-02    0.840161
 2029-12-31    0.822042
 Name: Zero Prices: 304, dtype: float64,
 2025-10-31    0.979291
 2026-04-30    0.958846
 2026-11-02    0.938948
 2027-04-30    0.919132
 2027-11-01    0.899442
 2028-05-01    0.879359
 2028-10-31    0.859813
 2029-04-30    0.840726
 2029-10-31    0.822358
 Name: Zero Prices: 304, dtype: float64,
 2025-07-31    0.979138
 2026-02-02    0.958690
 2026-07-31    0.938935
 2027-02-01    0.919015
 2027-08-02    0.899549
 2028-01-31    0.879982
 2028-07-31    0.860395
 2029-01-31    0.841154
 2029-07-31    0.822730
 2030-01-31    0.804344
 Name: Zero Prices: 304, dtype: float64]

### Calculate par yields for all the dates and combine with concat method of Pandas

In [15]:
# par yields for each bond's forward price series
dec_29=(1-forward_zero_prices[0])/forward_zero_prices[0].cumsum()*2
oct_29=(1-forward_zero_prices[1])/forward_zero_prices[1].cumsum()*2
jan_30=(1-forward_zero_prices[2])/forward_zero_prices[2].cumsum()*2

# combine and sort
par_yields=pd.concat([dec_29,oct_29,jan_30]).sort_index()

# display sorted results
display(par_yields)

2025-07-31    0.042613
2025-10-31    0.042295
2025-12-31    0.042335
2026-02-02    0.042635
2026-04-30    0.042467
2026-06-30    0.042456
2026-07-31    0.042454
2026-11-02    0.042440
2026-12-31    0.042447
2027-02-01    0.042671
2027-04-30    0.042605
2027-06-30    0.042694
2027-08-02    0.042788
2027-11-01    0.042830
2028-01-03    0.043052
2028-01-31    0.043053
2028-05-01    0.043279
2028-06-30    0.043400
2028-07-31    0.043385
2028-10-31    0.043571
2029-01-02    0.043788
2029-01-31    0.043658
2029-04-30    0.043783
2029-07-02    0.043952
2029-07-31    0.043773
2029-10-31    0.043874
2029-12-31    0.043965
2030-01-31    0.043948
Name: Zero Prices: 304, dtype: float64